url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/TimerQueue.java

```diff
@@ -0,0 +1,417 @@
+                        }
+                        if (m_sortedTimers.firstKey() == fireTime)
+                            m_cond.signal();
+                    }
+                    else
+                    {
+                        if (s_logger.isLoggable(Level.FINER))
+                        {
+                            s_logger.log( Level.FINER, System.identityHashCode(timerInfo.task) +
+                                    ": interval=" + interval );
+                        }
+                        timerInfo.next = next;
+                        next.prev = timerInfo;
+                    }
+                }
+            }
+            else
+            {
+                if (s_logger.isLoggable(Level.FINER))
+                    s_logger.log( Level.FINER, System.identityHashCode(timerInfo.task) + ": done" );
+                m_timers.remove( timerInfo.task );
+            }
+        }
+        finally
+        {
+            m_lock.unlock();
+        }
+
+        if (snatchThread)
+            runInThreadPool();
+    }
+
+    public void runInThreadPool()
+    {
+        if (s_logger.isLoggable(Level.FINE))
+            s_logger.log( Level.FINE, "started" );
+
+        m_lock.lock();
+        try
+        {
+            while (!m_sortedTimers.isEmpty())
+            {
+                final Map.Entry<Long, TimerInfo> firstEntry = m_sortedTimers.firstEntry();
+                assert( firstEntry != null );
+
+                final long currentTime = System.currentTimeMillis();
+                if (firstEntry.getKey() <= currentTime)
+                {
+                    if (s_logger.isLoggable(Level.FINER))
+                        s_logger.log( Level.FINER, "fireTime=" + firstEntry.getKey() + ": execute" );
+
+                    TimerInfo timerInfo = firstEntry.getValue();
+                    do
+                    {
+                        assert( timerInfo.threadID == 0 );
+                        final TimerInfo next = timerInfo.next;
+                        timerInfo.prev = null;
+                        timerInfo.next = null;
+                        timerInfo.threadID = -1; /* timer is being fired */
+
+                        m_threadPool.execute( timerInfo );
+                        timerInfo = next;
+                    }
+                    while (timerInfo != null);
+                    m_sortedTimers.remove( firstEntry.getKey() );
+                }
+                else
+                {
+                    final long sleepTime = (firstEntry.getKey() - currentTime);
+                    if (s_logger.isLoggable(Level.FINER))
+                        s_logger.log( Level.FINER, "firstEntry=" + firstEntry.getKey() + ", sleepTime=" + sleepTime );
+
+                    try
+                    {
+                        m_cond.awaitNanos( TimeUnit.MILLISECONDS.toNanos(sleepTime) );
+                    }
+                    catch (final InterruptedException ex)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)